File: sorting_algorithms/merge_sort.py
def merge(arr, low, middle, high):
    l = low
    h = middle

    result = []

    while l < middle and h < high:
        if arr[l] < arr[h]:
            result.append(arr[l])
            l += 1
        else:
            result.append(arr[h])
            h += 1
        
    if l < middle:
        result.extend(arr[l:middle])
    
    if h < high:
        result.extend(arr[h:high])
    
    for i in range(low, high):
        arr[i] = result[i - low]


def split(arr, low, high):
    middle = (low + high) // 2
    
    if low == middle:
        return

    split(arr, middle, high)
    split(arr, low, middle)
    merge(arr, low, middle, high)

def merge_sort(arr, low=0, high=None):
    if high is None:
        high = len(arr)
    
    split(arr, low, high)

    return arr

File: c/sortingAlgorithms/merge_sort.c
#include "utils.c"


void merge(int arr[], int low, int middle, int high) {
    int l = low;
    int h = middle;
    int index = 0;
    
    int result[LENGTH] = {0};

    while (l < middle && h < high) {
        if (arr[l] > arr[h]) {
            result[index] = arr[h];
            h++;
        } else {
            result[index] = arr[l];
            l++;
        }
        index++;
    }

    while (l < middle) {
        result[index] = arr[l];
        index++;
        l++;
    }
    
    while (h < high) {
        result[index] = arr[h];
        index++;
        h++;
    }

    for (int i = low; i < high; i++) {
        arr[i] = result[i-low];
    }
}

void merge_sort(int arr[], int low, int high) {
    int middle = (int) (low + high) / 2;
    
    if (low == middle) return;

    merge_sort(arr, middle, high);
    merge_sort(arr, low, middle);
    
    merge(arr, low, middle, high);
}

int main()
{
    print_array(arr, LENGTH);
    merge_sort(arr, 0, LENGTH);
    
    print_array(arr, LENGTH);
    printf("%s", equal_arrays(arr, sol, LENGTH) ? "true" : "false");

    return 0;
}